homeassistant/components/modbus/climate.py

```diff
@@ -156,13 +161,24 @@ async def async_set_temperature(self, **kwargs):
         target_temperature = int(
             (kwargs.get(ATTR_TEMPERATURE) - self._offset) / self._scale
         )
-        byte_string = struct.pack(self._structure, target_temperature)
-        struct_string = f">{DEFAULT_STRUCT_FORMAT[self._data_type]}"
-        register_value = struct.unpack(struct_string, byte_string)[0]
+        if self._data_type in [
+            DATA_TYPE_INT16,
+            DATA_TYPE_INT32,
+            DATA_TYPE_INT64,
+            DATA_TYPE_UINT16,
+            DATA_TYPE_UINT32,
+            DATA_TYPE_UINT64,
+        ]:
+            target_temperature = int(target_temperature)
+        else:
+            target_temperature = float(target_temperature)
+        byte_array = struct.pack(self._structure, target_temperature)
+        raw_regs = [byte_array[i : i + 2] for i in range(0, len(byte_array), 2)]
+        registers = self._swap_registers(raw_regs)
```

Comment:
I pulled this branch and tested locally, raw_regs is an array of bytearray, rather than ints. This is then cast to an int inside of _swap_registers() only if self._swap is `CONF_SWAP_BYTE` or `CONF_SWAP_WORD_BYTE`. In my case, not. This ultimately calls pymodbus with `registers` not being integer values.

Here is the log output for `count`=2 and `data_type`=`float32`:
```
2021-07-12 17:30:40 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [140373531464848] required argument is not an integer
Traceback (most recent call last):
  File "/workspaces/home-assistant/homeassistant/components/websocket_api/commands.py", line 185, in handle_call_service
    await hass.services.async_call(
  File "/workspaces/home-assistant/homeassistant/core.py", line 1491, in async_call
    task.result()
  File "/workspaces/home-assistant/homeassistant/core.py", line 1526, in _execute_service
    await handler.job.target(service_call)
  File "/workspaces/home-assistant/homeassistant/helpers/entity_component.py", line 213, in handle_service
    await self.hass.helpers.service.entity_service_call(
  File "/workspaces/home-assistant/homeassistant/helpers/service.py", line 658, in entity_service_call
    future.result()  # pop exception if have
  File "/workspaces/home-assistant/homeassistant/helpers/entity.py", line 811, in async_request_call
    await coro
  File "/workspaces/home-assistant/homeassistant/helpers/service.py", line 695, in _handle_entity_call
    await result
  File "/workspaces/home-assistant/homeassistant/components/climate/__init__.py", line 581, in async_service_temperature_set
    await entity.async_set_temperature(**kwargs)
  File "/workspaces/home-assistant/homeassistant/components/modbus/climate.py", line 181, in async_set_temperature
    result = await self._hub.async_pymodbus_call(
  File "/workspaces/home-assistant/homeassistant/components/modbus/modbus.py", line 318, in async_pymodbus_call
    result = await self.hass.async_add_executor_job(
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/workspaces/home-assistant/homeassistant/components/modbus/modbus.py", line 299, in _pymodbus_call
    result = self._pb_call[use_call][ENTRY_FUNC](address, value, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/pymodbus/client/common.py", line 103, in write_registers
    return self.execute(request)
  File "/usr/local/lib/python3.9/site-packages/pymodbus/client/sync.py", line 109, in execute
    return self.transaction.execute(request)
  File "/usr/local/lib/python3.9/site-packages/pymodbus/transaction.py", line 174, in execute
    response, last_exception = self._transact(
  File "/usr/local/lib/python3.9/site-packages/pymodbus/transaction.py", line 270, in _transact
    packet = self.client.framer.buildPacket(packet)
  File "/usr/local/lib/python3.9/site-packages/pymodbus/framer/socket_framer.py", line 206, in buildPacket
    data = message.encode()
  File "/usr/local/lib/python3.9/site-packages/pymodbus/register_write_message.py", line 164, in encode
    packet += struct.pack('>H', value)
struct.error: required argument is not an integer
```

So I suggest having _swap_registers only return int values. That should solve the int/bytearray type problem.

Reply:
hmmm seems you system works a bit different than my Mac, I will add a patch a bit later. Thanks for the review.

Reply:
The error was less obvious than I thought, but it is solved, and furthermore I have added tests to secure it works.